### PR TITLE
kdc101 control

### LINKED ISSUE #1: Add KDC101 control
Add support for [Thorlabs KCube KDC101 ](https://www.thorlabs.com/thorproduct.cfm?partnumber=KDC101)control.

- [x] Add unimplemented KCube messages to [MGMSG enum](https://github.com/quantumion/thorlabs_cube/blob/8bdb359f71e3ebd665259a74a0e7778d2e0139aa/thorlabs_tcube/driver.py#L12)
- [x] ~~Implement _Kcube base class?~~
- [x] Implement KDC101 class
- [x] Implement KDC101Sim class

### PR DESCRIPTION
**Pull request by collineps**
Add support for [KDC101 controller](https://www.thorlabs.com/thorproduct.cfm?partnumber=KDC101) and some package/repository management:

- Rename and refactor code for more modularity
- Add controller and simulation classes
- Add kdc101 as an option to the entrypoint script interface
- Fix interface to the asyncserial library
- Fix dependencies in setup script
- Add autodocumentation tools to optional dependencies
- Add KDC101 and TDC001 classes to documentation
- Enhance documentation with type hints
- Add protocol manual for reference
- Add a readme
- Add a .gitignore file

Control class successfully tested on hardware. Closes #1.

### CODE REVIEW

File: thorlabs_cube/driver/tcube/tdc.py [hunk 1]
```diff
@@ -0,0 +1,721 @@
...
+            vel1, wnd1, vel2, wnd2, vel3, wnd3, vel4. See
+            :py:meth:`set_pot_parameters()<Tdc.set_pot_parameters>` for a
+            description of each tuple element meaning.
+        :rtype: An 8 int tuple
+        """
+        get_msg = await self.send_request(MGMSG.MOT_REQ_POTPARAMS,
+                                          [MGMSG.MOT_GET_POTPARAMS], 1)
+        return st.unpack("<HLHLHLHL", get_msg.data[2:])
+
+    async def hub_get_bay_used(self):
```
Comment (kaonashi255) [inline thorlabs_cube/driver/tcube/tdc.py:78]: Doc string?

File: thorlabs_cube/driver/tcube/tpz.py [hunk 2]
```diff
@@ -0,0 +1,512 @@
...
+class Tpz(_Cube):
+    """Either :py:meth:`set_tpz_io_settings()<Tpz.set_tpz_io_settings>`
+    or :py:meth:`get_tpz_io_settings()<Tpz.get_tpz_io_settings>` must
+    be completed to finish initialising the driver.
+    """
+    def __init__(self, loop, serial_dev):
+        _Cube.__init__(self, loop, serial_dev)
+        self.voltage_limit = None
+
+    async def handle_message(self, msg):
```
Comment (kaonashi255) [inline thorlabs_cube/driver/tcube/tpz.py:16]: Doc string here or in `_Cube`?

**Comment by collineps** [comment]
The missing docstrings are in code from the original package that I didn't want to mess with. I can add some documentation from the protocol manual, but wanted to avoid messing with code unnecessarily. Think it's worth it?

File: doc/APT_Communications_Protocol_v38.pdf
[the remark this answers is not in the document]
Comment (collineps) [reply doc/APT_Communications_Protocol_v38.pdf:1]: I suppose it's not needed here, but a link depends on the vendor not removing the manual or changing to a version that's not compatible with the protocol used in this package.

File: thorlabs_cube/driver/kcube/kdc.py
[the remark this answers is not in the document]
Comment (collineps) [reply thorlabs_cube/driver/kcube/kdc.py:1]: Yep, pulled straight from the protocol manual.

File: doc/APT_Communications_Protocol_v38.pdf
[the remark this answers is not in the document]
Comment (kaonashi255) [reply doc/APT_Communications_Protocol_v38.pdf:1]: Okay, let's leave it here for now.

**Comment by kaonashi255** [comment]
> The missing docstrings are in code from the original package that I didn't want to mess with. I can add some documentation from the protocol manual, but wanted to avoid messing with code unnecessarily. Think it's worth it?

I guess not. Let's leave it as is.

**Review by kaonashi255 — APPROVED** [review]
The only "requested change" left is a question of style and isn't important.  I'm happy with everything else.

**Comment by collineps** [comment]
Ok sounds good. I'll open a documentation task to fill in the rest of the docstrings later.

File: thorlabs_cube/driver/kcube/kdc.py [hunk 3]
```diff
@@ -0,0 +1,289 @@
...
+        if msg_id == MGMSG.HW_DISCONNECT:
+            raise MsgError("Error: Please disconnect the KDC101")
+        elif msg_id == MGMSG.HW_RESPONSE:
+            raise MsgError("Hardware error, please disconnect and reconnect the KDC101")
+        elif msg_id == MGMSG.HW_RICHRESPONSE:
+            (code, ) = st.unpack("<H", data[2:4])
+            raise MsgError(f"Hardware error {code}: {data[4:].decode(encoding='ascii')}")
+        elif (msg_id == MGMSG.MOT_MOVE_COMPLETED or
+              msg_id == MGMSG.MOT_MOVE_STOPPED or
+              msg_id == MGMSG.MOT_GET_DCSTATUSUPDATE):
```
[the remark this answers is not in the document]
Comment (collineps) [reply thorlabs_cube/driver/kcube/kdc.py:35]: yes i think so. will make that change.

**Pull request merged**